core/src/main/scala/org/apache/spark/util/UninterruptibleThread.scala

```diff
@@ -101,7 +101,11 @@ private[spark] class UninterruptibleThread(
      *         is true) and no concurrent [[interrupt()]] call, otherwise false
      */
     def isInterruptible: Boolean = synchronized {
-      shouldInterruptThread = uninterruptible
+      // SPARK-53394: We should not interrupt the thread when it is already interrupted.
+      // Otherwise, the state of `shouldInterruptThread` becomes inconsistent between
+      // `isInterruptible()` and `isInterruptPending()`, leading to `UninterruptibleThread`
+      // be interruptible under `runUninterruptibly`.
+      shouldInterruptThread = uninterruptible || UninterruptibleThread.this.isInterrupted
```

Comment:
I don't think it is necessary to use `UninterruptibleThread.this` and you may use `isInterrupted`.

Reply:
We can't. Because we are in the inner class `UninterruptibleLock` rather than `UninterruptibleThread`.

Reply:
In Scala you can directly access outer class methods without using `Outer.this` (that would be necessary in Java). Can you give it a try (I checked and it compiles and runs)?